tests/phpunit/Presentation/TabsHtmlBuilderUnitTest.php

```diff
@@ -39,50 +41,31 @@ public function testTabsViewModelContainsItemTypeProperty(): void {
 
 	private function newTabsHtmlBuilder(
 		?Config $config = null,
+		?ConfigAuthorizer $configAuthorizer = null,
 		?SpyTemplateParser $templateSpy = null,
 		?QueryStringParser $queryStringParser = null,
 		?User $user = null
 	): TabsHtmlBuilder {
 		return new TabsHtmlBuilder(
-			$config ?? $this->newConfig(),
+			$config ?? new Config(),
+			$configAuthorizer ?? $this->newConfigAuthorizer(),
 			new FakeItemTypeLabelLookup(),
 			$templateSpy ?? new SpyTemplateParser(),
+			$titleFactory ?? MediaWikiServices::getInstance()->getTitleFactory(),
 			$queryStringParser ?? new StubQueryStringParser(),
-			$user ?? $this->newUser()
+			$user ?? $this->createMock( User::class )
 		);
 	}
 
-	private function newConfig( ?string $json = null ): Config {
-		if ( $json === null ) {
-			$json = <<<JSON
-{
-	"itemTypeProperty": "P1337",
-	"configPerItemType": {
-		"Q5976445": {
-			"facets": {
-				"P592": {
-					"type": "list"
-				}
-			}
-		}
-	}
-}
-JSON;
-		}
-
-		return WikibaseFacetedSearchExtension::getInstance()->newConfigDeserializer()->deserialize( $json );
-	}
-
-	private function newUser( bool $canEditConfig = false ): User {
-		$user = $this->createMock( User::class );
-		$user->method( 'isAllowed' )
-			->with( 'editsitejson' )
+	private function newConfigAuthorizer( bool $canEditConfig = false ): ConfigAuthorizer {
+		$configAuthorizer = $this->createMock( ConfigAuthorizer::class );
+		$configAuthorizer->method( 'isAuthorized' )
 			->willReturn( $canEditConfig );
-		return $user;
+		return $configAuthorizer;
```

Comment:
We generally try to avoid mocks. When you need a test double, you can change `ConfigAuthorizer` to an interface, and the current `ConfigAuthorizer` becomes something like `UserBasedConfigAuthorizer`. Then you create a test double like `SucceedingConfigAuthorizer` that always returns true.

Reply:
https://www.entropywins.wtf/blog/2016/05/13/5-ways-to-write-better-mocks/